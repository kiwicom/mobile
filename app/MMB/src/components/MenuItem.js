// @flow

import * as React from 'react';
import { View } from 'react-native';
import { Color, TextIcon, StyleSheet, Text } from '@kiwicom/mobile-shared';
import { Translation } from '@kiwicom/mobile-localization';

type Props = {|
  title: React.Element<typeof Translation>,
  icon: React.Element<typeof TextIcon>,
  description?: React.Element<typeof Translation>,
|};

/**
 * With optional description:
 *
 * .------------------------------.
 * |                              |
 * |  IC    Bold Title            |
 * |  ON    Optional description  |
 * |                              |
 * `------------------------------`
 *
 * Without description:
 *
 * .------------------------------.
 * |                              |
 * |  ICON  Bold Title            |
 * |                              |
 * `------------------------------`
 */
export default function MenuItem(props: Props) {
  return (
    <View style={styleSheet.wrapper}>
      <View style={styleSheet.icon}>{props.icon}</View>
      <View style={styleSheet.contentRow}>
        <Text style={styleSheet.title}>{props.title}</Text>
        {props.description && (
          <Text style={styleSheet.description}>{props.description}</Text>
        )}
      </View>
    </View>
  );
}

const styleSheet = StyleSheet.create({
  wrapper: {
    backgroundColor: Color.white,
    flexDirection: 'row',
    paddingVertical: 13,
    paddingHorizontal: 15,
  },
  icon: {
    paddingRight: 15,
    justifyContent: 'center',
  },
  contentRow: {},
  title: {
    fontWeight: 'bold',
  },
  description: {
    fontSize: 12,
    color: Color.textLight,
  },
});
