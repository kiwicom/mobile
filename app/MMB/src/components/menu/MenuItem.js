// @flow

import * as React from 'react';
import { View } from 'react-native';
import { Color, TextIcon, StyleSheet, Text } from '@kiwicom/mobile-shared';
import { Translation } from '@kiwicom/mobile-localization';

type Props = {|
  title: React.Element<typeof Translation>,
  icon?: React.Element<typeof TextIcon>,
  description?: React.Element<typeof Translation>,
|};

/**
 * With optional description:
 *
 * .------------------------------.
 * |                              |
 * |  IC    Bold Title            |
 * |  ON    Optional description  |
 * |                              |
 * `------------------------------`
 *
 * Without description (icon is optional as well):
 *
 * .------------------------------.
 * |                              |
 * |  ICON  Bold Title            |
 * |                              |
 * `------------------------------`
 */
export default function MenuItem(props: Props) {
  return (
    <View
      style={[
        styleSheet.wrapper,
        props.description ? styleSheet.wrapperWithDescription : null,
      ]}
    >
      {props.icon && <View style={styleSheet.icon}>{props.icon}</View>}

      <View style={styleSheet.contentRow}>
        <Text style={styleSheet.title}>{props.title}</Text>
        {props.description && (
          <Text style={styleSheet.description}>{props.description}</Text>
        )}
      </View>
    </View>
  );
}

const styleSheet = StyleSheet.create({
  wrapper: {
    backgroundColor: Color.white,
    flexDirection: 'row',
    paddingVertical: 20,
    paddingHorizontal: 15,
  },
  wrapperWithDescription: {
    paddingVertical: 13,
  },
  icon: {
    paddingRight: 15,
    justifyContent: 'center',
  },
  contentRow: {},
  title: {
    fontWeight: 'bold',
  },
  description: {
    fontSize: 12,
    color: Color.textLight,
  },
});
