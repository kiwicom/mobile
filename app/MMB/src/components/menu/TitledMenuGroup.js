// @flow

import * as React from 'react';
import { Platform } from 'react-native';
import { Translation } from '@kiwicom/mobile-localization';

import MenuGroup from './MenuGroup';
import MenuGroupTitle from './MenuGroupTitle';

type Props = {|
  title: React.Element<typeof Translation>,
  children: React.Node,
|};

/**
 * This component wraps <MenuItem/> and adds group title and visual separators.
 * Example:
 *
 *
 *   iOS Title
 * .------------------------------.
 * |                              |
 * |  ICON  Title 1               |
 * |                              |
 * |        ----------------------|
 * |                              |
 * |  ICON  Title 2               |
 * |                              |
 * `------------------------------`
 *
 *
 * .------------------------------.
 * |  Android Title               |
 * |------------------------------|
 * |                              |
 * |  ICON  Title 1               |
 * |                              |
 * |------------------------------|
 * |                              |
 * |  ICON  Title 2               |
 * |                              |
 * `------------------------------`
 */
export default function TitledMenuGroup({ title, children }: Props) {
  let newChildren = Array.isArray(children) ? children : [children];

  if (Platform.OS === 'android') {
    // we have to copy every children in order to prepend group title for Android
    newChildren = [<MenuGroupTitle key="android-ftw" title={title} />];
    React.Children.forEach(children, child => newChildren.push(child));
  }

  return (
    <React.Fragment>
      {Platform.OS === 'ios' && <MenuGroupTitle title={title} />}
      <MenuGroup withoutIcons={false}>{newChildren}</MenuGroup>
    </React.Fragment>
  );
}
