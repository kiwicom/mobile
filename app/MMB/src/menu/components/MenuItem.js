// @flow

import * as React from 'react';
import { View } from 'react-native';
import {
  Color,
  TextIcon,
  StyleSheet,
  Text,
  Touchable,
} from '@kiwicom/mobile-shared';
import { Translation } from '@kiwicom/mobile-localization';

type Props = {|
  title: React.Element<typeof Translation>,
  onPress: () => void,
  icon?: React.Element<typeof TextIcon>,
  description?: React.Element<typeof Translation>,
|};

/**
 * With optional description:
 *
 * .------------------------------.
 * |                              |
 * |  IC    Bold Title            |
 * |  ON    Optional description  |
 * |                              |
 * `------------------------------`
 *
 * Without description (icon is optional as well):
 *
 * .------------------------------.
 * |                              |
 * |  ICON  Bold Title            |
 * |                              |
 * `------------------------------`
 */
export default function MenuItem(props: Props) {
  return (
    <Touchable
      onPress={props.onPress}
      style={[
        styleSheet.wrapper,
        props.description ? styleSheet.wrapperWithDescription : null,
      ]}
    >
      <React.Fragment>
        {props.icon && <View style={styleSheet.icon}>{props.icon}</View>}

        <View style={styleSheet.contentRow}>
          <Text style={styleSheet.title}>{props.title}</Text>
          {props.description && (
            <Text style={styleSheet.description}>{props.description}</Text>
          )}
        </View>
      </React.Fragment>
    </Touchable>
  );
}

const styleSheet = StyleSheet.create({
  wrapper: {
    backgroundColor: Color.white,
    flexDirection: 'row',
    paddingVertical: 20,
    paddingHorizontal: 15,
  },
  wrapperWithDescription: {
    paddingVertical: 13,
  },
  icon: {
    paddingRight: 15,
    justifyContent: 'center',
  },
  contentRow: {},
  title: {
    fontWeight: 'bold',
  },
  description: {
    fontSize: 12,
    color: Color.textLight,
  },
});
