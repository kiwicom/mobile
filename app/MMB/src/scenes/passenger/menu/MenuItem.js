// @flow

import * as React from 'react';
import { View } from 'react-native';
import { Color, StyleSheet, Text } from '@kiwicom/mobile-shared';
import { Translation } from '@kiwicom/mobile-localization';

type Props = {|
  title: React.Element<typeof Translation>,
  description: React.Element<typeof Translation>,
|};

/**
 * This is basically a copy of /components/MenuItem
 * But it has a light title instead of bold
 * Example:
 *
 * .------------------------------.
 * |                              |
 * |  IC    Light Title           |
 * |  ON    Description           |
 * |                              |
 * `------------------------------`
 *
 */
export default function MenuItem(props: Props) {
  return (
    <View style={styles.wrapper}>
      <View style={styles.middleWrapper}>
        <Text style={styles.title}>{props.title}</Text>
        {props.description}
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    padding: 15,
    android: {
      backgroundColor: Color.white,
      borderStartWidth: 5,
      borderStartColor: Color.white,
    },
    ios: {
      backgroundColor: Color.white,
    },
  },
  middleWrapper: {
    flex: 1,
    justifyContent: 'center',
  },
  title: {
    color: Color.textLight,
  },
});
