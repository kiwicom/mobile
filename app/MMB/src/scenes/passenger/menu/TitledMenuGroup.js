// @flow

import * as React from 'react';
import { Platform } from 'react-native';
import { Translation } from '@kiwicom/mobile-localization';

import MenuGroup from './MenuGroup';
import MenuGroupTitle from '../../../components/menu/MenuGroupTitle';

type Props = {|
  title: React.Element<typeof Translation>,
  children: React.Node,
|};

// This is basically a copy of /components/TitledMenuGroup
// But it needs a different type of separator, so it uses ad different MenuGroup to wrap its children
export default function TitledMenuGroup({ title, children }: Props) {
  let newChildren = Array.isArray(children) ? children : [children];

  if (Platform.OS === 'android') {
    // we have to copy every children in order to prepend group title for Android
    newChildren = [<MenuGroupTitle key="android-ftw" title={title} />];
    React.Children.forEach(children, child => newChildren.push(child));
  }

  return (
    <React.Fragment>
      {Platform.OS === 'ios' && <MenuGroupTitle title={title} />}
      <MenuGroup>{newChildren}</MenuGroup>
    </React.Fragment>
  );
}
