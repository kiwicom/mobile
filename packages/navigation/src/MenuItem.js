// @flow

import * as React from 'react';
import { View } from 'react-native';
import {
  Color,
  TextIcon,
  StyleSheet,
  Text,
  Touchable,
  AdaptableLayout,
  Icon,
} from '@kiwicom/mobile-shared';
import { Translation } from '@kiwicom/mobile-localization';

import MenuItemIcon from './MenuItemIcon';

type Props = {|
  title: React.Element<typeof Translation>,
  onPress: () => void,
  isActive: boolean,

  // icon on the left
  icon?: React.Element<typeof TextIcon>,

  // icon on the right (arrow by default)
  actionIcon?: React.Element<typeof TextIcon>,
  description?: React.Element<typeof Translation>,
  testID?: string,
|};

/**
 * With optional description:
 *
 * .------------------------------.
 * |                              |
 * |  IC    Bold Title            |
 * |  ON    Optional description  |
 * |                              |
 * `------------------------------`
 *
 * Without description (icon is optional as well):
 *
 * .------------------------------.
 * |                              |
 * |  ICON  Bold Title            |
 * |                              |
 * `------------------------------`
 *
 * Menu item can be activated. Activation style differs with device:
 *
 *  - iOS: background color is brand and all text/icon colors are inverted
 *  - Android: only left box border is in the brand color
 */
function MenuItem(props: Props) {
  const styleSheet = createStyleSheet(props);

  return (
    <Touchable onPress={props.onPress} style={styleSheet.wrapper}>
      <React.Fragment>
        {props.icon && (
          <MenuItemIcon
            iconComponent={props.icon}
            invertColors={props.isActive}
          />
        )}

        <View style={styleSheet.middleWrapper}>
          <Text style={styleSheet.title}>{props.title}</Text>
          {props.description && (
            <Text style={styleSheet.description}>{props.description}</Text>
          )}
        </View>

        {props.actionIcon ? (
          <View style={styleSheet.rightArrow}>{props.actionIcon}</View>
        ) : (
          <AdaptableLayout
            renderOnNarrow={
              <View style={styleSheet.rightArrow}>
                <Icon name="chevron-right" size={26} color={Color.brand} />
              </View>
            }
          />
        )}
      </React.Fragment>
    </Touchable>
  );
}

AdaptableMenuItem.defaultProps = {
  isActive: false,
};

export default function AdaptableMenuItem(props: Props) {
  // Menu item can be activated but only on tablets (wide devices). Otherwise
  // it's not possible to activate it because new scene should always open
  // on mobile devices.
  return (
    <AdaptableLayout
      renderOnNarrow={<MenuItem {...props} isActive={false} />}
      renderOnWide={<MenuItem {...props} />}
    />
  );
}

function createStyleSheet(props: Props) {
  return StyleSheet.create({
    wrapper: {
      flexDirection: 'row',
      paddingVertical: 20,
      paddingHorizontal: props.description ? 13 : 15,
      android: {
        backgroundColor: Color.white,
        borderStartWidth: 5,
        borderStartColor: props.isActive ? Color.brand : Color.white,
      },
      ios: {
        backgroundColor: props.isActive ? Color.brand : Color.white,
      },
    },
    middleWrapper: {
      flex: 1,
      justifyContent: 'center',
    },
    title: {
      fontWeight: 'bold',
      android: {
        color: Color.textDark,
      },
      ios: {
        color: props.isActive ? Color.white : Color.textDark,
      },
    },
    description: {
      fontSize: 12,
      android: {
        color: Color.textLight,
      },
      ios: {
        color: props.isActive ? Color.white : Color.textLight,
      },
    },
    rightArrow: {
      justifyContent: 'center',
    },
  });
}
