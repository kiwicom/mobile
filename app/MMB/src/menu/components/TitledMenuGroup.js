// @flow

import * as React from 'react';
import { View } from 'react-native';
import { Color, StyleSheet, Text } from '@kiwicom/mobile-shared';
import { Translation } from '@kiwicom/mobile-localization';

import MenuItem from './MenuItem';
import MenuGroup from './MenuGroup';

type Props = {|
  title: React.Element<typeof Translation>,
  children: $ReadOnlyArray<React.Element<typeof MenuItem>>,
|};

/**
 * This component wraps <MenuItem/> and adds group title and visual separators.
 * Example:
 *
 *
 *   Trip Info Separator
 * .------------------------------.
 * |                              |
 * |  ICON  Title 1               |
 * |                              |
 * |        ----------------------|
 * |                              |
 * |  ICON  Title 2               |
 * |                              |
 * `------------------------------`
 */
export default function TitledMenuGroup(props: Props) {
  return (
    <React.Fragment>
      <View style={styleSheet.titleWrapper}>
        <Text style={styleSheet.title}>{props.title}</Text>
      </View>

      <MenuGroup withoutIcons={false}>{props.children}</MenuGroup>
    </React.Fragment>
  );
}

const styleSheet = StyleSheet.create({
  titleWrapper: {
    paddingTop: 22,
    paddingBottom: 11,
    paddingLeft: 15,
  },
  title: {
    color: Color.textLight,
  },
});
