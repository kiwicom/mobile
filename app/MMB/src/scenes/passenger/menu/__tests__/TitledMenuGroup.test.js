// @flow

import * as React from 'react';
import { PlaygroundRenderer } from '@kiwicom/mobile-playground';
import { Translation } from '@kiwicom/mobile-localization';

import TitledMenuGroup from '../TitledMenuGroup';
import MenuItem from '../MenuItem';

describe('TitledMenuGroup', () => {
  it('renders', () => {
    PlaygroundRenderer.render(
      <TitledMenuGroup title={<Translation passThrough="GroupTitle!" />}>
        <MenuItem
          title={<Translation passThrough="Title!" />}
          description={<Translation passThrough="Description!" />}
        />
        <MenuItem
          title={<Translation passThrough="Title!" />}
          description={<Translation passThrough="Description!" />}
        />
      </TitledMenuGroup>,
      false,
      'PassengerTitledMenuGroup',
    );
  });
});
