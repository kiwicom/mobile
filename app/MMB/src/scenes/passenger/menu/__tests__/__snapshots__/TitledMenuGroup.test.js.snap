// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitledMenuGroup renders 1`] = `
<UNDEFINED>
  <MenuGroupTitle
    title={
      <Translation
        passThrough="GroupTitle!"
      />
    }
  />
  <MenuGroup>
    <MenuItem
      description={
        <Translation
          passThrough="Description!"
        />
      }
      title={
        <Translation
          passThrough="Title!"
        />
      }
    />
    <MenuItem
      description={
        <Translation
          passThrough="Description!"
        />
      }
      title={
        <Translation
          passThrough="Title!"
        />
      }
    />
  </MenuGroup>
</UNDEFINED>
`;
